File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    # 돗자리를 큰 것부터 시도하기 위해 내림차순 정렬
    mats.sort(reverse=True)
    rows = len(park)
    cols = len(park[0])

    # 특정 크기의 돗자리를 해당 위치에 깔 수 있는지 확인하는 함수
    def can_place(mat_size, r, c):
        if r + mat_size > rows or c + mat_size > cols:
            return False  # 돗자리가 공원 범위를 벗어남

        for i in range(r, r + mat_size):
            for j in range(c, c + mat_size):
                if park[i][j] != "-1":
                    return False  # 사람이나 장애물이 있는 경우 배치 불가
        return True

    # 가장 큰 돗자리부터 확인
    for mat in mats:
        for i in range(rows):
            for j in range(cols):
                if can_place(mat, i, j):
                    return mat  # 가장 큰 돗자리를 찾으면 즉시 반환

    return -1  # 배치할 수 있는 돗자리가 없으면 -1 반환

File: 프로그래머스/1/340000/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
import java.util.Arrays;

class Solution {
    public int solution(int[] mats, String[][] park) {
        // 돗자리를 큰 것부터 시도하기 위해 내림차순 정렬
        Arrays.sort(mats);
        reverseArray(mats);

        int rows = park.length;
        int cols = park[0].length;

        // 특정 크기의 돗자리를 해당 위치에 깔 수 있는지 확인하는 메소드
        for (int mat : mats) {
            for (int i = 0; i <= rows - mat; i++) {
                for (int j = 0; j <= cols - mat; j++) {
                    if (canPlace(mat, i, j, park)) {
                        return mat; // 가장 큰 돗자리를 찾으면 즉시 반환
                    }
                }
            }
        }
        return -1; // 배치할 수 있는 돗자리가 없으면 -1 반환
    }

    // 돗자리를 해당 위치에 배치할 수 있는지 확인하는 메소드
    private boolean canPlace(int matSize, int r, int c, String[][] park) {
        for (int i = r; i < r + matSize; i++) {
            for (int j = c; j < c + matSize; j++) {
                if (!park[i][j].equals("-1")) {
                    return false; // 사람이 있거나 장애물이 있는 경우 배치 불가
                }
            }
        }
        return true;
    }

    // 배열을 내림차순으로 정렬하는 메소드
    private void reverseArray(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }
}
